[java] PMD crashes with ClassFormatError: Absent Code attribute...

**Affects PMD Version:** 6.18.0

**Rule:** n/a

**Description:**
PMD crashes with the following stacktrace:

```
Caused by: java.lang.ClassFormatError: Absent Code attribute in method that is not native or abstract in class file javax/xml/ws/Service
    at java.lang.ClassLoader.defineClass1 (Native Method)
    at java.lang.ClassLoader.defineClass (ClassLoader.java:1016)
    at java.security.SecureClassLoader.defineClass (SecureClassLoader.java:151)
    at java.net.URLClassLoader.defineClass (URLClassLoader.java:515)
    at java.net.URLClassLoader$1.run (URLClassLoader.java:423)
    at java.net.URLClassLoader$1.run (URLClassLoader.java:417)
    at java.security.AccessController.doPrivileged (AccessController.java:689)
    at java.net.URLClassLoader.findClass (URLClassLoader.java:416)
    at net.sourceforge.pmd.util.ClasspathClassLoader.loadClass (ClasspathClassLoader.java:114)
    at java.lang.ClassLoader.loadClass (ClassLoader.java:521)
    at java.lang.ClassLoader.defineClass1 (Native Method)
    at java.lang.ClassLoader.defineClass (ClassLoader.java:1016)
    at java.security.SecureClassLoader.defineClass (SecureClassLoader.java:151)
    at java.net.URLClassLoader.defineClass (URLClassLoader.java:515)
    at java.net.URLClassLoader$1.run (URLClassLoader.java:423)
    at java.net.URLClassLoader$1.run (URLClassLoader.java:417)
    at java.security.AccessController.doPrivileged (AccessController.java:689)
    at java.net.URLClassLoader.findClass (URLClassLoader.java:416)
    at net.sourceforge.pmd.util.ClasspathClassLoader.loadClass (ClasspathClassLoader.java:114)
    at java.lang.ClassLoader.loadClass (ClassLoader.java:521)
    at java.lang.Class.getDeclaringClass0 (Native Method)
    at java.lang.Class.getEnclosingClass (Class.java:1522)
    at net.sourceforge.pmd.lang.java.typeresolution.typedefinition.JavaTypeDefinitionSimple.<init> (JavaTypeDefinitionSimple.java:68)
    at net.sourceforge.pmd.lang.java.typeresolution.typedefinition.JavaTypeDefinition.forClass (JavaTypeDefinition.java:74)
    at net.sourceforge.pmd.lang.java.ast.AbstractAnyTypeDeclaration.setQualifiedName (AbstractAnyTypeDeclaration.java:90)
    at net.sourceforge.pmd.lang.java.qname.QualifiedNameResolver.visit (QualifiedNameResolver.java:212)
    at net.sourceforge.pmd.lang.java.ast.JavaParserVisitorReducedAdapter.visit (JavaParserVisitorReducedAdapter.java:16)
    at net.sourceforge.pmd.lang.java.ast.ASTClassOrInterfaceDeclaration.jjtAccept (ASTClassOrInterfaceDeclaration.java:55)
    at net.sourceforge.pmd.lang.java.ast.AbstractJavaNode.childrenAccept (AbstractJavaNode.java:61)
    at net.sourceforge.pmd.lang.java.ast.JavaParserVisitorAdapter.visit (JavaParserVisitorAdapter.java:11)
    at net.sourceforge.pmd.lang.java.ast.JavaParserVisitorAdapter.visit (JavaParserVisitorAdapter.java:56)
    at net.sourceforge.pmd.lang.java.ast.ASTClassOrInterfaceBodyDeclaration.jjtAccept (ASTClassOrInterfaceBodyDeclaration.java:44)
    at net.sourceforge.pmd.lang.java.ast.AbstractJavaNode.childrenAccept (AbstractJavaNode.java:61)
    at net.sourceforge.pmd.lang.java.ast.JavaParserVisitorAdapter.visit (JavaParserVisitorAdapter.java:11)
    at net.sourceforge.pmd.lang.java.ast.JavaParserVisitorAdapter.visit (JavaParserVisitorAdapter.java:51)
    at net.sourceforge.pmd.lang.java.ast.ASTClassOrInterfaceBody.jjtAccept (ASTClassOrInterfaceBody.java:35)
    at net.sourceforge.pmd.lang.java.ast.AbstractJavaNode.childrenAccept (AbstractJavaNode.java:61)
    at net.sourceforge.pmd.lang.java.ast.JavaParserVisitorAdapter.visit (JavaParserVisitorAdapter.java:11)
    at net.sourceforge.pmd.lang.java.ast.JavaParserVisitorReducedAdapter.visit (JavaParserVisitorReducedAdapter.java:33)
    at net.sourceforge.pmd.lang.java.qname.QualifiedNameResolver.visit (QualifiedNameResolver.java:214)
    at net.sourceforge.pmd.lang.java.ast.JavaParserVisitorReducedAdapter.visit (JavaParserVisitorReducedAdapter.java:16)
    at net.sourceforge.pmd.lang.java.ast.ASTClassOrInterfaceDeclaration.jjtAccept (ASTClassOrInterfaceDeclaration.java:55)
    at net.sourceforge.pmd.lang.java.ast.AbstractJavaNode.childrenAccept (AbstractJavaNode.java:61)
    at net.sourceforge.pmd.lang.java.ast.JavaParserVisitorAdapter.visit (JavaParserVisitorAdapter.java:11)
    at net.sourceforge.pmd.lang.java.ast.JavaParserVisitorAdapter.visit (JavaParserVisitorAdapter.java:206)
    at net.sourceforge.pmd.lang.java.ast.ASTTypeDeclaration.jjtAccept (ASTTypeDeclaration.java:39)
    at net.sourceforge.pmd.lang.java.ast.AbstractJavaNode.childrenAccept (AbstractJavaNode.java:61)
    at net.sourceforge.pmd.lang.java.ast.JavaParserVisitorAdapter.visit (JavaParserVisitorAdapter.java:11)
    at net.sourceforge.pmd.lang.java.ast.JavaParserVisitorAdapter.visit (JavaParserVisitorAdapter.java:181)
    at net.sourceforge.pmd.lang.java.qname.QualifiedNameResolver.visit (QualifiedNameResolver.java:134)
    at net.sourceforge.pmd.lang.java.ast.ASTCompilationUnit.jjtAccept (ASTCompilationUnit.java:44)
    at net.sourceforge.pmd.lang.java.qname.QualifiedNameResolver.initializeWith (QualifiedNameResolver.java:118)
    at net.sourceforge.pmd.lang.java.AbstractJavaHandler$8.start (AbstractJavaHandler.java:153)
    at net.sourceforge.pmd.SourceCodeProcessor.resolveQualifiedNames (SourceCodeProcessor.java:125)
    at net.sourceforge.pmd.SourceCodeProcessor.processSource (SourceCodeProcessor.java:176)
```


**Code Sample demonstrating the issue:**
Thanks to @gmshake the issue can be reproduced like that: https://github.com/adangel/scratchpad/pull/1


**Running PMD through:** Maven

--- a/docs/pages/release_notes.md
+++ b/docs/pages/release_notes.md
@@ -16,6 +16,8 @@ This is a {{ site.pmd.release_type }} release.
 
 ### Fixed Issues
 
+*   java
+    *   [#2042](https://github.com/pmd/pmd/issues/2042): \[java] PMD crashes with ClassFormatError: Absent Code attribute...
 *   java-codestyle
     *    [#2017](https://github.com/pmd/pmd/issues/2017): \[java] UnnecessaryFullyQualifiedName triggered for inner class
 

--- a/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/typeresolution/typedefinition/JavaTypeDefinitionSimple.java
+++ b/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/typeresolution/typedefinition/JavaTypeDefinitionSimple.java
@@ -65,7 +65,15 @@ import java.util.logging.Logger;
             this.genericArgs = NO_GENERICS;
         }
 
-        enclosingClass = forClass(clazz.getEnclosingClass());
+        Class<?> enclosing = null;
+        try {
+            enclosing = clazz.getEnclosingClass();
+        } catch (LinkageError e) {
+            if (LOG.isLoggable(Level.WARNING)) {
+                LOG.log(Level.WARNING, "Could not load enclosing class of " + clazz.getName() + ", due to: " + e);
+            }
+        }
+        enclosingClass = forClass(enclosing);
     }
 
     @Override

